src/Common/src/CoreWCF/Runtime/NonReentrantAsyncLock.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CoreWCF.Runtime
+{
+    /// <summary>
+    /// This class provides a locking mechanism compatible with task async patterns.
+    /// </summary>
+    /// <remarks>
+    /// Unlike the <see cref="ReentrantAsyncLock"/> it does not provide a reentrant (aka. recursive) locking
+    /// and can therefore only be used in scenarios where we have clear control over the execution path.
+    /// </remarks>
+    internal sealed class NonReentrantAsyncLock
+    {
+        // based of the ideas in https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-6-asynclock/
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+        private readonly Task<IDisposable> _releaser;
+#if DEBUG
+        private static readonly AsyncLocal<bool> s_reentranceFlag = new AsyncLocal<bool>();
```

Comment:
This won't work. If you have 2 NonReentrantAsyncLock instances being used in the same call path, when Lock A is acquired, it will set the reentranceFlag, and when Lock B is acquired, it will think it's already been acquired. You need the AsyncLocal to be an instance variable. But then this is where you have problems with memory leaks. Once you set an AsyncLocal to a value, it gets stored in the ExecutionContext. If that ExecutionContext lives for a long time (such as can happen with timers), then you end up growing and growing the ExecutionContext. The solution is to set the AsyncLocal value to null, but you can't do that with a bool. When you set it to null, it actually removes it's storage slot from the ExecutionContext completely. The exception to this is if you add a change notification handler, then it stores that it's a null and will still leak. The fix is to make this an instance variable of type `AsyncLocal<object>`. You set the value to a new object instance when the lock is acquired, and you set it to null when it's released. Your Assert check validates that the Value is null.